cmd/cp-url.go

```diff
@@ -158,20 +176,62 @@ func makeCopyContentTypeB(sourceAlias string, sourceContent *ClientContent, targ
 
 // SINGLE SOURCE - Type C: copy(d1..., d2) -> []copy(d1/f, d1/d2/f) -> []A
 // prepareCopyRecursiveURLTypeC - prepares target and source clientURLs for copying.
-func prepareCopyURLsTypeC(ctx context.Context, sourceURL, targetURL string, isRecursive, isZip bool, timeRef time.Time) <-chan URLs {
+func prepareCopyURLsTypeC(ctx context.Context, sourceContent, targetContent *ClientContent, sourceURL, targetURL string, encKeyDB map[string][]prefixSSEPair, isRecursive, isZip bool, timeRef time.Time) <-chan URLs {
 	// Extract alias before fiddling with the clientURL.
 	sourceAlias, _, _ := mustExpandAlias(sourceURL)
 	// Find alias and expanded clientURL.
 	targetAlias, targetURL, _ := mustExpandAlias(targetURL)
-	copyURLsCh := make(chan URLs)
-	go func(sourceURL, targetURL string, copyURLsCh chan URLs) {
-		defer close(copyURLsCh)
-		sourceClient, err := newClient(sourceURL)
+	copyURLsCh := make(chan URLs, 1)
+
+	returnErrorAndCloseChannel := func(err *probe.Error) chan URLs {
+		copyURLsCh <- URLs{Error: err}
+		close(copyURLsCh)
+		return copyURLsCh
+	}
+
+	c, err := newClient(sourceURL)
+	if err != nil {
+		return returnErrorAndCloseChannel(err.Trace(sourceURL))
+	}
+
+	if targetContent == nil {
+		_, targetContent, err = url2Stat(ctx, targetURL, "", false, encKeyDB, time.Time{}, isZip)
 		if err != nil {
-			// Source initialization failed.
-			copyURLsCh <- URLs{Error: err.Trace(sourceURL)}
-			return
+			return returnErrorAndCloseChannel(err.Trace(targetURL))
 		}
+	}
+
+	if !targetContent.Type.IsDir() {
+		return returnErrorAndCloseChannel(errTargetIsNotDir(targetURL).Trace(targetURL))
+	}
+
+	if sourceContent == nil {
+		_, sourceContent, err = url2Stat(ctx, sourceURL, "", false, encKeyDB, time.Time{}, isZip)
+		if err != nil {
+			return returnErrorAndCloseChannel(err.Trace(sourceURL))
+		}
+	}
+
+	if sourceContent.Type.IsDir() {
+		// Require --recursive flag if we are copying a directory
+		if !isRecursive {
+			return returnErrorAndCloseChannel(errRequiresRecursive(sourceURL).Trace(sourceURL))
+		}
+
+		// Check if we are going to copy a directory into itself
+		if isURLContains(sourceURL, targetURL, string(c.GetURL().Separator)) {
+			return returnErrorAndCloseChannel(errCopyIntoSelf(sourceURL).Trace(targetURL))
+		}
+	}
+
+	// if !isRecursive {
+	// 	copyURLsCh <- makeCopyContentTypeC(sourceAlias, c.GetURL(), sourceContent, targetAlias, targetURL)
+	// 	close(copyURLsCh)
+	// 	return copyURLsCh
+	// }
```

Comment:
Remove if no longer needed.

Reply:
done, it was a test for a fix I'll be working on next. Just forgot to remove it.